[doc] add description of common switches (#5484)
openthread/openthread: examples/README.md

﻿# Common switches

> OpenThread allows you to [configure](https://openthread.io/guides/build#configuration) the stack for different functionality and behavior. This configuration is based on changing compile-time constants during the build process using common switches listed in `/examples/common-switches.mk`.

Reviewer: Nit: I would probably note somewhere that all the switches default to 0 unless otherwise noted in the description.

> OpenThread allows you to [configure](https://openthread.io/guides/build?hl=en#configuration) the stack for different functionality and behavior. This configuration is based on changing compile-time constants during the build process using common switches listed in `/examples/common-switches.mk`.

Reviewer: You can drop the ?hl=en from the configure link

> This page lists the available common switches with description and additional information. For build command examples, see [build examples](https://openthread.io/guides/build?hl=en#build_examples).

Reviewer: Drop ?hl=en from this link too

> | Makefile switch | CMake switch | Default value | Description | Additional information |

Reviewer: Thanks for putting this together @piotrkoziar, great to finally have a reference of all the switches.  I have a major suggestion for the format: the table is a bit wide because some flags are long and requires horizontal scrolling to view, we should try to eliminate that for readability.  I would suggest a few things:

1. Combine the Description and Additional information columns, the Description is often repeated in the Additional Info, and Additional Info is blank half the time anyway
1. Drop the "Default value" column since nearly all of them default to 0... instead note at the top that all switches default to 0 unless otherwise noted (which we could do in the Description column)
1. Don't spell out a full link like https://openthread.io/guides/build/features/child-supervision... link the inline text directly

If you don't have the time to reformat, we can do it in a later pass, but I think condensing the columns would work well.

Alternatively, you can try adding `<wbr>` tags in the longer switch strings to force line breaks and reduce the column widths.

| --- | --- | --- | --- | --- |

> | BACKBONE_ROUTER | OT_BACKBONE_ROUTER | 0 | Enables backbone router functionality for Thread 1.2. |  |

Reviewer 2: ```suggestion
| BACKBONE_ROUTER | OT_BACKBONE_ROUTER | 0 | Enables Backbone Router functionality for Thread 1.2. |  |
```

| BIG_ENDIAN | OT_BIG_ENDIAN | 0 | Allows the host platform to use big-endian byte order. |  |
| BORDER_AGENT | OT_BORDER_AGENT | 0 | Enables support for border agent. | In most cases, enable this switch if you are building On-mesh Commissioner or Border Router with External Commissioning support. |

> | BORDER_ROUTER | OT_BORDER_ROUTER | 0 | Enables support for border router. | This switch is usually combined with the BORDER_AGENT and UDP_FORWARD switches to build Border Router device. |

Reviewer 2: But for the `RCP` design, we don't use `UDP_FORWARD` but use `PLATFORM_UDP`.

> | BORDER_ROUTER | OT_BORDER_ROUTER | 0 | Enables support for border router. | This switch is usually combined with the BORDER_AGENT and UDP_FORWARD switches to build Border Router device. |

Reviewer 2: ```suggestion
| BORDER_ROUTER | OT_BORDER_ROUTER | 0 | Enables support for Border Router. | This switch is usually combined with the BORDER_AGENT and UDP_FORWARD switches to build Border Router device. |
```

| CHANNEL_MANAGER | OT_CHANNEL_MANAGER | 0 | Enables support for channel manager. | Enable this switch on devices that are supposed to request a Thread network channel change. This switch should be used only with an FTD build. |
| CHANNEL_MONITOR | OT_CHANNEL_MONITOR | 0 | Enables support for channel monitor. | Enable this switch on devices that are supposed to determine the cleaner channels. |
| COAP | OT_COAP | 0 | Enables support for the CoAP API. | Enable this switch if you want to control Constrained Application Protocol communication. |
| COAPS | OT_COAPS | 0 | Enables support for the secure CoAP API. | Enable this switch if you want to control Constrained Application Protocol Secure (CoAP over DTLS) communication. |
| COAP_OBSERVE | OT_COAP_OBSERVE | 0 | Enables support for CoAP Observe (RFC7641) API. |  |

> | COMMISSIONER | OT_COMMISSIONER | 0 | Enables support for commissioner. | Enable this switch on device that is able to perform Commissioner role. CMake swtich: . |

Reviewer 2: ```suggestion
| COMMISSIONER | OT_COMMISSIONER | 0 | Enables support for Commissioner. | Enable this switch on device that is able to perform Commissioner role. CMake swtich: . |
```

> | COVERAGE | not implemented | 0 | Enables the generation of code-coverage instances. |  |

Reviewer 2: ```suggestion
| COVERAGE | OT_COVERAGE | 0 | Enables the generation of code-coverage instances. |  |
```

> | CHILD_SUPERVISION | OT_CHILD_SUPERVISION | 0 | Enables support for child supervision. | Enable this switch on a parent or child node with custom OpenThread application that manages the supervision, checks timeout intervals, and verifies connectivity between parent and child. Read more about the child supervision: https://openthread.io/guides/build/features/child-supervision. |

Reviewer: I would avoid having the link completely spelled out to ensure the column width isn't too big, instead link the text directly:

See [Child Supervision](https://openthread.io/guides/build/features/child-supervision) to learn more.

| CLI_TRANSPORT | not implemented | UART | Selects the transport of CLI. | You can set this switch to UART (default) or CONSOLE. |

> | CSL_RECEIVER | OT_CSL_RECEIVER | 0 | Enables CSL receiver. |  |

Reviewer 2: ```suggestion
| CSL_RECEIVER | OT_CSL_RECEIVER | 0 | Enables CSL receiver feature for Thread 1.2. |  |
```

| DEBUG | not implemented | 0 | Allows building debug instance. Code optimization is disabled. |  |
| DHCP6_CLIENT | OT_DHCP6_CLIENT | 0 | Enables support for the DHCP6 client. | The device is able to act as typical DHCP client. Enable this switch on a device that is supposed to request networking parameters from the DHCP server. |
| DHCP6_SERVER | OT_DHCP6_SERVER | 0 | Enables support for the DHCP6 server. | The device is able to act as typical DHCP server. Enable this switch on a device that is supposed to provide networking parameters to devices with DHCP_CLIENT switch enabled. |
| DIAGNOSTIC | OT_DIAGNOSTIC | 0 | Enables diagnostic support. | Enable this switch on a device that is tested in the factory production stage. |
| DISABLE_DOC | not implemented | 0 | Disables building of the documentation. |  |
| DISABLE_TOOLS | not implemented | 0 | Disables building of tools. |  |
| DNS_CLIENT | OT_DNS_CLIENT | 0 | Enables support for DNS client. | Enable this switch on a device that sends a DNS query for AAAA (IPv6) record. |
| DUA | OT_DUA | 0 | Enables the Domain Unicast Address feature for Thread 1.2. |  |

> | DYNAMIC_LOG_LEVEL | not implemented | 0 | Enables the dynamic log level feature. | Enable this switch if OpenThread log level is required to be set at runtime. Read more about logging: https://openthread.io/reference/group/api-logging. |

Reviewer: Would it make more sense to link to the Logging guide rather than the reference?  For this and other related switches like LOG_OUTPUT, REFERENCE_DEVICE, FULL_LOGS etc.

https://openthread.io/guides/build/logs

| ECDSA | OT_ECDSA | 0 | Enables support for Elliptic Curve Digital Signature Algorithm. | Enable this switch if ECDSA digital signature is used by application. |

> | EXTERNAL_HEAP | OT_EXTERNAL_HEAP | 0 | Enables support for external heap. | Make sure to specify the external heap Calloc and Free functions to be used by the OpenThread stack. Enable this switch if the platform uses its own heap. |

Reviewer: Seems like the "Enable this switch if the platform uses its own heap." sentence should be first.

| IP6_FRAGM | OT_IP6_FRAGM | 0 | Enables support for IPv6 fragmentation. |  |
| JAM_DETECTION | OT_JAM_DETECTION | 0 | Enables support for Jam Detection. | Enable this switch if a device requires the ability to detect signal jamming on a specific channel. Read more about Jam Detection: https://openthread.io/guides/build/features/jam-detection. |

> | JOINER | OT_JOINER | 0 | Enables support for joiner. | Enable this switch on a device that has to be commissioned to join the network. Read more about the joiner API: https://openthread.io/reference/group/api-joiner. |

Reviewer 2: ```suggestion
| JOINER | OT_JOINER | 0 | Enables support for Joiner. | Enable this switch on a device that has to be commissioned to join the network. Read more about the Joiner API: https://openthread.io/reference/group/api-joiner. |
```

| LEGACY | OT_LEGACY | 0 | Enables support for legacy network. |  |
| LINK_RAW | OT_LINK_RAW | 0 | Enables the Link Raw service. |  |
| LOG_OUTPUT | not implemented | PLATFORM_DEFINED | Defines if the LOG output is to be created and where it goes. | There are several options available: `NONE`, `DEBUG_UART`, `APP`, `PLATFORM_DEFINED` (default). |
| MAC_FILTER | OT_MAC_FILTER | 0 | Enables support for the MAC filter. |  |

> | MLE_LONG_ROUTES | OT_MLE_LONG_ROUTES | 0 | Enables the MLE long routes extension (experimental, breaks Thread conformance). |  |

Reviewer: Is "Thread conformance" the same as Thread Specification?

Reviewer 2: by `breaks Thread conformance` I meant that turning on this feature breaks Thread Specification conformance.

Reviewer: Let's make it consistent across switches then, as TIME_SYNC states:

**Note: this feature is incompatible with Thread Specification.**

Which I assume is the same thing.  We should say something like "Enabling this feature breaks conformance to the Thread Specification." for both

| MLR | OT_MLR | 0 | Enables Multicast Listener Registration feature for Thread 1.2. |  |
| MTD_NETDIAG | OT_MTD_NETDIAG | 0 | Enables the TMF network diagnostics on MTDs. |  |
| MULTIPLE_INSTANCE | OT_MULTIPLE_INSTANCE | 0 | Enables multiple OpenThread instances. |  |
| PLATFORM_UDP | OT_PLATFORM_UDP | 0 | Enables platform UDP support. |  |
| REFERENCE_DEVICE | OT_REFERENCE_DEVICE | 0 | Enables support for Thread Test Harness reference device. | Enable this switch on the reference device during certification. |
| SERVICE | OT_SERVICE | 0 | Enables support for injecting Service entries into the Thread Network Data. |  |
| SETTINGS_RAM | OT_SETTINGS_RAM | 0 | Enables volatile-only storage of settings. |  |
| SLAAC | OT_SLAAC | 1 | Enables support for adding auto-configured SLAAC addresses by OpenThread. |  |
| SNTP_CLIENT | OT_SNTP_CLIENT | 0 | Enables support for SNTP Client. |  |

> | THREAD_VERSION | OT_THREAD_VERSION | 1.1 | Enables the chosen Thread version. | For example, set to `1.1` for Thread v1.1. |

Reviewer 2: ```suggestion
| THREAD_VERSION | OT_THREAD_VERSION | 1.1 | Enables the chosen Thread version (1.1 / 1.2). | For example, set to `1.1` for Thread 1.1. |
```

> | THREAD_VERSION | OT_THREAD_VERSION | 1.1 | Enables the chosen Thread version (1.1 / 1.2). | For example, set to `1.1` for Thread 1.1. |

Reviewer: Since the default is 1.1, I would use 1.2 in the example

> | TIME_SYNC | OT_TIME_SYNC | 0 | Enables the time synchronization service feature. |  |

Reviewer 2: ```suggestion
| TIME_SYNC | OT_TIME_SYNC | 0 | Enables the time synchronization service feature. **Note: this feature is incompatible with Thread Specification.** |  |
```

| UDP_FORWARD | OT_UDP_FORWARD | 0 | Enables support for UDP forward. | Enable this switch on the Border Router device (running on the NCP design) with External Commissioning support to service Thread Commissioner packets on the NCP side. |
| DISABLE_BUILTIN_MBEDTLS | not implemented | 0 | Disables OpenThread's mbedTLS build. | Enable this switch if you do not want to use the built-in mbedTLS and you do not want to manage mbedTLS internally. Enabling this switch will disable support for such features as memory allocation and debug. |
| BUILTIN_MBEDTLS_MANAGEMENT | OT_BUILTIN_MBEDTLS_MANAGEMENT | 0 | Enables the built-in mbedTLS management. | Enable this switch if the external mbedTLS is used, but mbedTLS memory allocation and debug config should be managed internally by OpenThread. |
| DISABLE_EXECUTABLE | not implemented | 0 | Disables building of executables. |  |
| DEBUG_UART | not implemented | 0 | Enables the Debug UART platform feature. |  |
| DEBUG_UART_LOG | not implemented | 0 | Enables the log output for the debug UART. | Requires OPENTHREAD_CONFIG_ENABLE_DEBUG_UART to be enabled. |

> | OTNS | OT_OTNS | 0 | Enables support for OpenThread Network Simulator. | Enable this switch if you are building OpenThread for OpenThread Network Simulator. |

Reviewer 2: ```suggestion
| OTNS | OT_OTNS | 0 | Enables support for [OpenThread Network Simulator](https://github.com/openthread/ot-ns). | Enable this switch if you are building OpenThread for OpenThread Network Simulator. |
```

| SPINEL_ENCRYPTER_LIBS | not implemented | '' | Specifies library files (absolute paths) for implementing the NCP Spinel Encrypter. |  |

> | FULL_LOGS | not implemented | 0 | Enables all log levels and regions. | This switch sets the log level to OT_LOG_LEVEL_DEBG and turns on all region flags. |

Reviewer 2: ```suggestion
| FULL_LOGS | OT_FULL_LOGS | 0 | Enables all log levels and regions. | This switch sets the log level to OT_LOG_LEVEL_DEBG and turns on all region flags. |
```

You can configure OpenThread for different functionality and behavior during the build process. Available configuration options are detailed at the following locations:

- Compile-time constants are listed in all the header files in /src/core/config
- Makefile build switches are listed in /examples/common-switches.mk

Use a switch to enable functionality for an example platform. For example, to build the nrf52840 example with Commissioner and Joiner support enabled:

```
make -f examples/Makefile-nrf52840 COMMISSIONER=1 JOINER=1
```

Or, to build the nRF52840 example with the Jam Detection feature enabled:

```
make -f examples/Makefile-nrf52840 JAM_DETECTION=1
```

The common-switches are described below.

| Makefile switch | Default value | Description | Additional information |
| --- | --- | --- | --- |

> | BACKBONE_ROUTER | 0 | Enable backbone router functionality. | CMake switch: OT_BACKBONE_ROUTER. |

Reviewer 2: Mark it as Thread 1.2 only?
